### ISSUE TITLE
odd byte sequences can confuse mdl

### ISSUE THREAD

**Issue by jaymzh** [markdownlint/markdownlint#278]
## Description
Byte sequences that sometimes trigger rules that then crash. For example, non-printable characters followed by `*` followed by non-printable characters trigger list-based rules, but then the rules don't make sense on those byte sequences.

Here are some specific sequences:
```
  "\x41\x0A\x0C\x0A\x2A\x20\x42",
  "\x41\x0D\x0C\x0D\x2A\x20\x42",
  "\x41\x0D\x0A\x0C\x0D\x0A\x42",
```

This bug was originally reported by @jamesquilty over in #237 bit it was a separate issue.

## Replication Case

You can generate test cases with:

```ruby
#!/usr/bin/ruby

byte_strings = [
  "\x41\x0A\x0C\x0A\x2A\x20\x42",
  "\x41\x0D\x0C\x0D\x2A\x20\x42",
  "\x41\x0D\x0A\x0C\x0D\x0A\x42",
]

byte_strings.each_with_index do |bytes, index|
  file = "/tmp/wonk#{index + 1}"
  puts "Generating #{file}"
  open(file, 'w+b') do |fh|
    fh.write(bytes)
  end
end
```

And then run mdl on the associated files and watch it crash.

**Comment by jaymzh** [comment]
My analysis from the other bug:

So I wrote this script to generate test files for me:

```ruby
byte_strings = [
  "\x41\x0A\x0C\x0A\x2A\x20\x42",
  "\x41\x0D\x0C\x0D\x2A\x20\x42",
  "\x41\x0D\x0A\x0C\x0D\x0A\x42",
]

byte_strings.each_with_index do |bytes, index|
  file = "/tmp/wonk#{index + 1}"
  puts "Generating #{file}"
  open(file, 'w+b') do |fh|
    fh.write(bytes)
  end
end
```

I find that mdl, as it is, works fine on the 1st and 3rd strings and only crashes on the 2nd. That's on `master` with no modifications at all. Changing the split to use lines doesn't change this at all.

So the reason this fails is because once it parses it the "lines" it gets us:

```
["A\r\f\r* B"]
```

So that is something non-printable followed by a `*` followed by more non-printable. So **then** we get down to line list rule which tries to look the line number for this element and can't find it.

So I suspect the proper fix, might be something like this:

```
@@ -190,7 +191,12 @@ module MarkdownLint
       if item.type != :li
         raise "list_style called with non-list element"
       end
-      line = element_line(item).strip
+      line = element_line(item)
+      unless line
+        raise "list_style called with bogus data"
+      end
+
+      line.strip!
```

Which will still bail out, but gives a slightly clear error. Or possibly:

```
@@ -190,7 +191,14 @@ module MarkdownLint
       if item.type != :li
         raise "list_style called with non-list element"
       end
-      line = element_line(item).strip
+      line = element_line(item)
+      unless line
+        puts "Warning: Skipping potential list item, could not determine" +
+          " line: #{item}"
+        return :unknown
+      end
+      line.strip!
+
```

Which gets us further, but we then die trying to determine the indentation on the list a bit later. We can fake that out:

```
@@ -211,6 +219,7 @@ module MarkdownLint
     # indent of 8 spaces. You need to pass in the raw string here.
 
     def indent_for(line)
+      return 0 unless line
       return line.match(/^\s*/)[0].gsub("\t", " " * 8).length
     end
```

but we then crash later, and you see where that goes. I would argue the right solution is simply to throw a nicer error as-above.

**Comment by jaymzh** [comment]
OK it turns out once we fix the splitting per #238 the crashes all unify in a single place that's easy to handle. PR coming.

**Comment by jaymzh** [comment]
https://github.com/markdownlint/markdownlint/pull/279 will, after the other PR merges, prevent the errors... though I'm not 100% convinced it's the best solution. @apjanke ?

**Comment by jaymzh** [comment]
@jamesquilty - can you try the attached patch?
@apjanke - any thoughts?

**Comment by jamesquilty** [comment]
@jaymzh Thanks for writing a fix for this bug! I tried the 2-line patch from b5341f8 on version 0.5.0 and it doesn't crash on a saved test file, while the unpatched version does crash. I'm not sure that means too much, and I'd suggest you write an "odd byte sequence" CI test that could catch regressions - it will give feedback faster than I'm typically able :-/

**Comment by jaymzh** [comment]
Yeah I just don't love the silent "ignore whatever we just hit" aspect of this PR... basically if one part of the code thought we hit a list, but then another part thinks we didn't, we just move on. I'm somewhat tempted to add a special flag that people can pass in that says `--try-to-ignore-strange-things` or some such...

**Comment by jaymzh** [comment]
Fixed in 0.16.0
